query/aggregate.go

```diff
@@ -0,0 +1,210 @@
+	sum := 0.0
+	for _, v := range array {
+		sum += v
+	}
+	return sum
+}
+
+// A mean aggregator is highly similar to a sum aggregator.
+// It computes the aggregate mean for a seriesList
+type meanAggregator struct {
+}
+
+var _ aggregator = meanAggregator{}
+
+// The mean aggregator returns the mean of the given array
+func (aggregator meanAggregator) aggregate(array []float64) float64 {
+	if len(array) == 0 {
+		// The mean of an empty list is not well-defined
+		return math.NaN()
+	}
+	sum := 0.0
+	for _, v := range array {
+		sum += v
+	}
+	return sum / float64(len(array))
+}
+
+// The min aggregator is an aggregator that computes the aggregate minimum for a seriesList
+type minAggregator struct {
+}
+
+var _ aggregator = minAggregator{}
+
+func (aggregator minAggregator) aggregate(array []float64) float64 {
+	if len(array) == 0 {
+		// The minimum of an empty list is not well-defined
+		return math.NaN()
+	}
+	min := array[0]
+	for _, v := range array {
+		min = math.Min(min, v)
+	}
+	return min
+}
+
+// The maxAggregator is an aggregator that computes the aggregate maximum for a seriesList
+type maxAggregator struct {
+}
+
+var _ aggregator = maxAggregator{}
+
+func (aggregator maxAggregator) aggregate(array []float64) float64 {
+	if len(array) == 0 {
+		// The maximum of an empty list is not well-defined
+		return math.NaN()
+	}
+	max := array[0]
+	for _, v := range array {
+		max = math.Max(max, v)
+	}
+	return max
+}
+
+// applyAggregation takes an aggregation function ( [float64] => float64 ) and applies it to a given list of Timeseries
+// the list must be non-empty, or an error is returned
+func applyAggregation(group group, aggregator aggregator) (api.Timeseries, error) {
+	list := group.List
+	tagSet := group.TagSet
+
+	if len(list) == 0 {
+		return api.Timeseries{}, EmptyAggregateError{}
+	}
+
+	series := api.Timeseries{
+		Values: make([]float64, len(list[0].Values)), // The first Series in the given list is used to determine this length
+		TagSet: tagSet,                               // The tagset is supplied by an argument (it will be the values grouped on)
+	}
+
+	// Make a slice of time to reuse.
+	// Each entry corresponds to a particular Series, all having the same index within their corresponding Series.
+	timeSlice := make([]float64, len(list))
+
+	for i := range series.Values {
+		// We need to determine each value in turn.
+		for j := range timeSlice {
+			timeSlice[j] = list[j].Values[i]
+		}
+		// Find the aggregated value:
+		series.Values[i] = aggregator.aggregate(timeSlice)
+	}
+
+	return series, nil
+}
+
+// This function is the culmination of all others.
+// `aggregateBy` takes a series list, an aggregator, and a set of tags.
+// It produces a SeriesList which is the result of grouping by the tags and then aggregating each group
+// into a single Series.
+func aggregateBy(list api.SeriesList, aggregator aggregator, tags []string) (api.SeriesList, error) {
```

Comment:
We should have callers use an enum instead of an actual `aggregator` when calling into this. `aggregators` end up being just an `aggregate()`, so we can key the enum into a hash of `aggregate()` functions.